File: algorithms/quick_sort/ruby/quick_sort.rb
def quicksort(array)
  return array if array.length <= 1

  pivot_index = (array.length / 2).to_i
  pivot_value = array[pivot_index]
  array.delete_at(pivot_index)

  lesser = Array.new
  greater = Array.new

  array.each do |x|
    if x <= pivot_value
      lesser << x
    else
      greater << x
    end
  end

  quicksort(lesser) + [pivot_value] + quicksort(greater)
end

arr = [1,8,4,5,7,3,9,2,6]
quicksort(arr)
# => [1, 2, 3, 4, 5, 6, 7, 8, 9]

File: algorithms/quick_sort/javascript/quick_sort.js
/* Quicksort implementation in JavaScript */

function quicksort(arr, p, r) {
	if ( p < r ) {
		q = partition(arr, p, r);
		quicksort(arr, p, q-1);
		quicksort(arr, q+1, r);
	}
	return arr;
}


function partition(arr, p, r) {
	x = arr[r];
	i = p-1;
	var temp;
	for(var j = p; j < r; j++) {
		if (arr[j] < x) {
			i++;
			temp = arr[i];
			arr[i] = arr[j];
			arr[j] = temp;
		}
	}
	i++;
	temp = arr[i];
	arr[i] = arr[r];
	arr[r] = temp;

	return i;
}
	

var arr = [9, 8, 7, 6, 5, 4, 3, 2, 1];
console.log(quicksort(arr, 0, arr.length-1));
